## MetaPhase Software Package

Created by Josh Burton (`jnburton at uw.edu`) and Ivan Liachko (`il34 at uw.edu`) in the Department of Genome Sciences at the University of Washington, Seattle, WA, USA


## Source code

The `src/` directory contains the main source code for the computational portion of MetaPhase.  This is the software that performs the clustering algorithm that is central to the metagenomic deconvolution method.  Inputs include:

* A draft metagenome assembly
* A SAM/BAM file containing Hi-C read alignments to this assembly
* (Optional) A set of reference genomes of species that you suspect might be in your sample

The main output is a set of FASTA files representing the clusters of contigs.  The MetaPhase package can also produce output files to visualize the clustering result as a network (e.g., Fig. 2a, 3a in the paper), a validation heatmap (e.g., Fig. 2b, 3b in the paper), or an *.XGMML file suitable for loading into Cytoscape.

The fasta files generated by MetaPhase can be input into Lachesis to produce scaffolded assemblies of individual genomes.  For information on running Lachesis, see [the Lachesis publication](http://dx.doi.org/10.1038/nbt.2727) in *Nature Biotechnology* and [the Lachesis GitHub page](https://github.com/shendurelab/Lachesis).  Note that Lachesis is only recommended for assembly of eukaryotic genomes.


## Scripts

The `scripts/` directory contains several scripts that I used to prepare the datasets for the paper.  I've included them here as a way of documenting what I did for the MetaPhase paper.  I haven't modified these scripts at all, so they won't run as-is (because, for example, they've got directory names from my filesystem hard-coded into them.)

- `assemble.sh`: A wrapper script for [IDBA-UD](http://i.cs.hku.hk/~alse/hkubrg/projects/idba_ud/index.html), the _de novo_ metagenome assembler.  Used to create the M-Y (MetaYeast) draft assembly.
- `SplitFastaIntoBins.pl`: A script that splits all the contigs in a fasta file into bins of a specified size.  Used to create the M-3D simulated assembly.
- `FindContigsOnRefs.pl`: A script to determine which contigs in a draft assembly map to which of a set of reference genomes, and report summary stats.  Used to create the supplementary figure indicating draft contig coverage of the reference genomes.
- `blast.sh`: A wrapper script to call [blastn](http://blast.ncbi.nlm.nih.gov/Blast.cgi).  Used in `FindContigsOnRefs.pl`.
- `align.iter.sh`: An iterative aligner.  Trims reads, tries to align them with [BWA](http://bio-bwa.sourceforge.net/), then trims them again and repeats.
